llvm/test/Transforms/MemCpyOpt/callslot_deref.ll

```diff
@@ -6,7 +6,7 @@ declare void @llvm.memcpy.p0.p0.i64(ptr nocapture, ptr nocapture readonly, i64,
 declare void @llvm.memset.p0.i64(ptr nocapture, i8, i64, i1) nounwind
 
 ; all bytes of %dst that are touch by the memset are dereferenceable
-define void @must_remove_memcpy(ptr noalias nocapture dereferenceable(4096) %dst) nofree nosync {
+define void @must_remove_memcpy(ptr noalias nocapture writable dereferenceable(4096) %dst) nofree nosync {
```

Comment:
If `dereferenceable(N)` + `writable` means no-data race to write these pointers, are `noalias`s no longer required for those tests? Seems like the current implementation seems so.

Reply:
In this particular case it is redundant, yes (and also wasn't needed before this change). But in the general case it is still needed. E.g. if we have
```llvm
define void @test(ptr writable dereferenceable(128) %dst) {
  %src = alloca [128 x i8], align 4
  call void @accept_ptr(ptr nocapture %src) nounwind
  call void @llvm.memcpy.p0.p0.i64(ptr align 4 %dst, ptr %src, i64 128, i1 false)
  ret void
}
```
then we can't perform the transform because `@accept_ptr` could modify `%dst`. It depends on whether there are any ways to indirectly modify `%dst`.

Reply:
Ah, I see this is what AA matters. Thanks!